### ISSUE TITLE
[patterns] Can we make type declarations and type tests look different?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#2677]
I'm a little bit worried about patterns taking a step backwards, so to speak, in that they use a syntax which is otherwise declaring the type of a variable in a way that turns it into a dynamic type test. We're going to get a match failure rather than a dynamic error, but otherwise it's very similar to the implicit downcasts that we had a couple of years ago.

My main worries are (1) it is difficult to discern the actual semantics based on reading the source code, and (2) it's easy to get those type annotations wrong, and then they are just silently turned into type tests (this makes the construct similar to implicit downcasts, except that we're getting different/buggy control flow rather than getting an exception).

```dart
void main() {
  List<num> xs = [1, 2, 3];
  switch (xs) {
    case [Object o, num n, int i]: ...;
  }
}
```

The static type of `xs` is used to determine some elements of the list pattern (in particular, inferring the type argument `num`). The three variables are treated differently.

We might want to introduce a variable like `o` with a more general type because we might want to assign other values to `o` than the value that it gets via pattern matching. So it's not necessarily a mistake to give `o` the declared type `Object`, and we have to do it because the inferred type would otherwise be `num`, and then we couldn't assign those other values later on.

The variable `n` gets the type `num` that we would otherwise have gotten by inference. This may be useful for documentation purposes.

The variable `i` gets the specified type `int`, but pattern matching will now include a type test: The element at `[2]` which is known to be a `num` may or may not be an `int` as well, but we will test this at run time.

I think it would be helpful if we used a different syntax for the cases where the declared type is guaranteed to be satisfied by the matched value, and the cases where it may or may not hold, and there will be a type test at run time.

One possible approach could be to simply say that the version with a dynamic type test is not supported: If we declare a type which is not statically known to hold for the matched value then it's a compile-time error. The developer must then perform the dynamic type test in some other way (for instance, `when i is int`).

Another possibility would be to say that the type test can't be requested using the syntax `int i` (that's still an error), but you could use a different syntax, like `var i is int`. This would be similar to patterns using `!= null`, and it would provide the promotion based on syntax which is already recognizable as such.

If we do this then `num n` would actually be more useful: It would serve as a contract between `n` and the surrounding source code: I'm expecting `n` to be able to have the type `num`, and I'd like to get a heads-up if this is not the case.

(Note that this wouldn't change anything for object patterns: We can still match an object using a dynamic type test for `int` by matching it with the object pattern `int()`&mdash;it is unsurprising that this kind of matching includes a type test, because the ability to handle different cases corresponding to a set of distinct subtypes of the scrutinee type is absolutely standard for all pattern matching mechanisms.)

**Comment by lrhn** [comment]
The problem here is that we cannot distinguish an up-cast from a type check.
Which means we can't give the author a warning if an intended up-cast is actually a (probably failing) type check.
Using the same syntax for both prevents the author from showing intent.

We have the same problem with casts today, where you have to use an `as` cast for an up-cast in expression position. If you do it correctly, the cast should be compiled to a no-op, so you are wasting good syntax on doing nothing. If you make a mistake, it'll be a runtime error.

We are using an operation which can fail at runtime, to do something which we know cannot fail.

And that hurts when you make mistakes, because it becomes a run-time error.
That's true for
```dart
int x = foo;
var y = x as double; // whoops, meant `num`. And it works on the web too!
```
It's true for:
```dart
if (o case Object(:double hashCode)) ...
```
which will never work. (Except on the web, whoops.)

OK, I admit it's a real problem. I like the terseness of the current syntax, but it does mean you have to conflate different intentions into the same syntax.

Using a suffix `pattern is type` gets too wordy, `num name is int` is four *words* with no punctuation, only spaces. That's hard to read.

Prefix `is type` is slightly better, but not as general. We'd need `is int num x` to match an `int` and upcast to `num` in one test. Probably have to require `is int (num x)` for that, and only allow `is int x` to bind to the same type as matched.
Reasonable for the common case, upcasting after a type check is going to be very rare (why not just match `num`?).

I like it (and I'll say more below). I like using `is` instead of `case` for `if`-patterns. (I'd want it as a general expression to be used anywhere an `e is type` can be used today, but it'll probably need parentheses outside of very limited contexts).

The cost is more verbosity.
```dart
case (int x, int y): ...
```
becomes
```dart
case (is int x, is int y): ...
// or
case is (int, int) (var x, var y): ...
```
Unless we can do:
```dart
case is (int x, int y): ...
```
and infer the required type of `(int x, int y)` as the type to test for. (We totally should!)

So: 
* `is type (id | pattern)?` means check the value is the type `type`, then bind it to `id` with type `type`  or match it against `pattern` with matched value type `type`. (And `is type pattern` is just the same as `is type & pattern`).
* `is pattern` means: Let *S* be the type schema of pattern. (Find the matched value type if necessary.) Fill out any holes in the schema based on the matched value type, call it `T`.  Then treat as `is T (pattern)`. The pattern cannot be a type literal, since then it matches the first item instead.
* A plain typed binding pattern of `type id` must be valid against the matched value type. It's not a type check.
* An object pattern of the form `SomeType(...)` is also not a type check. You can want to extract properties from a supertype of the matched value type. With inline class types, or even just extension methods, that might change the available getters. So it should also require a leading `is`.
* A list or map pattern should probably be type checking inherently. It's rare that you'd want to up-cast to `List` or `Map` (but you may want to up-cast the type parameters, again because of non-virtual methods, so maybe those require `is` too, and compile-time failures if the cast from matched value type to the list/map type is not an upcast.)

It starts to feel like there will be entire switches where every case starts with `is`. Should we allow those to use `is` instead of `case`? So:
```dart
switch (e) {
  case 0: ....; // A constant
  is [int x, int y]: // ...  type schema is List<int>, so that's what we check for.
  is Foo(: var x, : var y): // ...
  case null: ....
  case _: .... // Binding with no type check, but also cannot be a type failure.
}
```

I don't know. It gets complicated. The ability to distinguish up-casts from type-check means that we have to be explicit about it in every pattern. Most patterns are not up-casting, so requiring the new/extra syntax for type checks might be going in the wrong direction.
But the failure mode of making a mistake is to do a type check, so it *should* be very you have to opt in.
It's ... complicated.


__About `is` in `if`-patterns:__
We don't need to switch to `is` in `if`-patterns, we can do everything the same with `case` instead.
Switching to `is` should be because it's *better*.
Allowing both `if (e is int)` and `if (e is int ix)` feels *nice* to me. It may not look much better than `case` for more complicated checks, but probably also not worse.

The main worry would be that `int` now means something different in `if (e is int)` and `switch (e) { case int:` (and `var int = e;`). The context-sensitivity still mainly applies to raw identifiers, but now also extends to entire types with type arguments (which can be constants - damn you, type literals!).
They would all mean the same thing as what they means today, which is a bonus, and we don't have to explain why it's `e is T` but `e case T x`.
And you can still switch/add context by doing `switch (e) case is int: ...`, `var (int _) = e;` or `is (int)`.

I like it that. Even if it's just allowing `e is type-pattern` in `if`-patterns. I think it can hold water. I'd like to explore it more! :)

For generally requiring `is` for any type check, so that we can catch mis-written upcasts, I fear it brings us too far from the syntactic sweet spot, and from other pattern languages.

**Comment by munificent** [comment]
OK, I spent some time thinking about this more. I'm definitely sympathetic with the desire to distinguish type annotations from type tests. But given how many patterns do type tests now, I think it would be a very large effort to change the proposal to make this distinction. I think it's very unlikely we could come up with a coherent, complete design and get all the issues with it shaken out in time.

Even if we could, it's not clear to me that doing so would be an improvement. The current proposal is fairly similar to how Scala and C# behave and it's seemed to work out well for those languages.

Unless others feels strongly, I'm inclined to say that we should leave the current proposal as it is. We've had a lot of eyeballs on it from the community, and this aspect isn't one that I can recall many raising issues with, so I suspect the current design is OK.

**Comment by munificent** [comment]
Closing this because we're going to stick with the current design.
